### PR TITLE
[WIP] Reading g_GCShadow directly in JIT_WriteBarrier() for arm64

### LINKED ISSUE #40058: [HeapVerify=2] Assert failure "Pointer updated without using write barrier"
Hi,

### Description

Testcase: [GC/API/NoGCRegion/NoGC/NoGC.exe](https://github.com/dotnet/runtime/tree/master/src/tests/GC/API/NoGCRegion)

```
export CORE_LIBRARIES=/home/zhaixiang/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.6.20264.1
export COMPlus_HeapVerify=2
/home/zhaixiang/runtime/artifacts/bin/coreclr/Linux.arm64.Debug/corerun /home/zhaixiang/coreclr-mips64-dev/bin/tests/Linux.arm64.Debug/GC/API/NoGCRegion/NoGC/NoGC.exe


Workstation on 64-bit with 4 procs
=====allocating 100mb allowing full blocking GC first=====

Calling TryStartNoGCRegion(..) with totalSize = 100 MB
100 MB SUCCEEDED, did 1 gen2 GCs
before GC: 980, after GC: 980
ended no gc region
current GC count: 1955

=====allocating 100mb allowing full blocking GC first Succeeded=====
=====allocating 200mb allowing full blocking GC first=====

Calling TryStartNoGCRegion(..) with totalSize = 200 MB
200 MB SUCCEEDED, did 1 gen2 GCs
before GC: 2938, after GC: 2938
ended no gc region

Assert failure(PID 20106 [0x00004e8a], Thread: 20106 [0x4e8a]): !"Pointer updated without using write barrier"
    File: /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp Line: 38307
    Image: /home/zhaixiang/runtime/artifacts/bin/coreclr/Linux.arm64.Debug/corerun

Aborted
```

### Configuration

* [coreclr-mips64-dev](https://github.com/gsvm/coreclr), [coreclr v3.1.6](https://github.com/dotnet/coreclr/tree/v3.1.6) and [runtime master](https://github.com/dotnet/runtime/commit/e671f57b1b8a6efc8febc023a59fdb0a76b5d286) are able to reproduce the issue.
* Loongnix 1.0 (docker image: aoqi/dotnet-buildtools:loongson3a-loongnix-1.0-llvm8ld) and Ubuntu 16.04.3 LTS
* ARM64 and MIPS64

### Other information

```
Assert failure(PID 20298 [0x00004f4a], Thread: 20298 [0x4f4a]): !"Pointer updated without using write barrier"
    File: /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp Line: 38307
    Image: /home/zhaixiang/runtime/artifacts/bin/coreclr/Linux.arm64.Debug/corerun


Thread 1 "corerun" received signal SIGTRAP, Trace/breakpoint trap.
DBG_DebugBreak () at /home/zhaixiang/runtime/src/coreclr/src/pal/src/arch/arm64/debugbreak.S:8
8	    EMIT_BREAKPOINT
(gdb) bt
#0  DBG_DebugBreak () at /home/zhaixiang/runtime/src/coreclr/src/pal/src/arch/arm64/debugbreak.S:8
#1  0x0000007fb74876d8 in DebugBreak () at /home/zhaixiang/runtime/src/coreclr/src/pal/src/debug/debug.cpp:405
#2  0x0000007fb6d5c878 in DbgAssertDialog (szFile=0x7fb77fc4f8 "/home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp", iLine=38307, szExpr=0x7fb78034df "!\"Pointer updated without using write barrier\"") at /home/zhaixiang/runtime/src/coreclr/src/utilcode/debug.cpp:698
#3  0x0000007fb72b7080 in WKS::testGCShadow (ptr=0x7f1800e4a8) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:38307
#4  WKS::testGCShadowHelper (x=0x7f1800ae78 "\220\205U>\177") at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:38324
#5  0x0000007fb7287e24 in WKS::checkGCWriteBarrier () at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:38352
#6  0x0000007fb7287644 in WKS::gc_heap::garbage_collect (n=0) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:17502
#7  0x0000007fb7270848 in WKS::GCHeap::GarbageCollectGeneration (this=0x55555c3780, gen=0, reason=reason_alloc_soh) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:36873
#8  0x0000007fb7272620 in WKS::gc_heap::trigger_gc_for_alloc (gen_number=0, gr=reason_alloc_soh, msl=0x7fb7b7ec48 <WKS::gc_heap::more_space_lock_soh>, loh_p=false, take_state=WKS::mt_try_budget) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:13336
#9  0x0000007fb7273994 in WKS::gc_heap::try_allocate_more_space (acontext=0x55555eb028, size=1984, flags=0, gen_number=0) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:13459
#10 0x0000007fb7273c78 in WKS::gc_heap::allocate_more_space (acontext=0x55555eb028, size=1984, flags=0, alloc_generation_number=0) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:13895
#11 0x0000007fb72b2790 in WKS::gc_heap::allocate (jsize=1977, acontext=0x55555eb028, flags=0) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:13926
#12 WKS::GCHeap::Alloc (this=0x55555c3780, context=0x55555eb028, size=1977, flags=0) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:35979
#13 0x0000007fb6f3ed28 in Alloc (size=1977, flags=GC_ALLOC_NO_FLAGS) at /home/zhaixiang/runtime/src/coreclr/src/vm/gchelpers.cpp:229
#14 0x0000007fb6f3d368 in AllocateSzArray (pArrayMT=0x7f3e50f0f0, cElements=1953, flags=GC_ALLOC_NO_FLAGS) at /home/zhaixiang/runtime/src/coreclr/src/vm/gchelpers.cpp:484
#15 0x0000007fb6f591f8 in JIT_NewArr1 (arrayMT=0x7f3e50f0f0, size=1953) at /home/zhaixiang/runtime/src/coreclr/src/vm/jithelpers.cpp:2718
#16 0x0000007fb6f58cc8 in JIT_NewArr1VC_MP_FastPortable (arrayMT=0x7f3e50f0f0, size=1953) at /home/zhaixiang/runtime/src/coreclr/src/vm/jithelpers.cpp:2621
#17 0x0000007f3e2d8e68 in ?? ()
#18 0x0000007f18014ad0 in ?? ()
(gdb) x/22i 0x0000007f3e2d8e68-44
   0x7f3e2d8e3c:	ldr	x2, [x2]
   0x7f3e2d8e40:	ldr	x2, [x2,#72]
   0x7f3e2d8e44:	ldr	x2, [x2,#56]
   0x7f3e2d8e48:	blr	x2
   0x7f3e2d8e4c:	str	w0, [x29,#60]
   0x7f3e2d8e50:	ldr	w1, [x29,#60]
   0x7f3e2d8e54:	sxtw	x1, w1
   0x7f3e2d8e58:	mov	x0, #0xf0f0                	// #61680
   0x7f3e2d8e5c:	movk	x0, #0x3e50, lsl #16
   0x7f3e2d8e60:	movk	x0, #0x7f, lsl #32
   0x7f3e2d8e64:	bl	0x7f3e29db10
   0x7f3e2d8e68:	str	x0, [x29,#48]
   0x7f3e2d8e6c:	ldr	x0, [x29,#48]
   0x7f3e2d8e70:	str	x0, [x29,#72]
   0x7f3e2d8e74:	ldr	w0, [x29,#84]
   0x7f3e2d8e78:	ldr	w1, [x29,#88]
   0x7f3e2d8e7c:	cmp	w1, #0x0
   0x7f3e2d8e80:	b.ne	0x7f3e2d8e88
   0x7f3e2d8e84:	bl	0x7f3e29daf0
   0x7f3e2d8e88:	cmn	w1, #0x1
   0x7f3e2d8e8c:	b.ne	0x7f3e2d8ea0
   0x7f3e2d8e90:	cmn	w0, w0
```

Thanks,
Leslie Zhai

### PR DESCRIPTION
**Pull request by cshung**
Fixes https://github.com/dotnet/runtime/issues/40058, see https://github.com/dotnet/runtime/issues/40058#issuecomment-680176146 for the investigation leading to this fix.

@dotnet/gc, @dotnet/jit-contrib

### CODE REVIEW

File: src/coreclr/src/vm/arm64/asmhelpers.S [hunk 1]
```diff
@@ -375,7 +375,7 @@ WRITE_BARRIER_ENTRY JIT_WriteBarrier
     // Update GC Shadow Heap
 
     // Do not perform the work if g_GCShadow is 0
-    ldr  x12, LOCAL_LABEL(wbs_GCShadow)
```
Comment (mangod9) [inline src/coreclr/src/vm/arm64/asmhelpers.S:378]: is `wbs_GCShadow` still required? Also is this similar to how its done for x64?

Comment (cshung) [reply src/coreclr/src/vm/arm64/asmhelpers.S:378]: Nope, we don't need the label anymore, I can remove it.

**Comment by echesakovMSFT** [comment]
As I understood `HeapVerify=2` infrastructure is working on arm64 now. Do you see any value to add such the corresponding test scenarios (e.g. GCStress+HeapVerify2 in runtime-coreclr gcstress-extra test pipeline) in the CI?

**Comment by mangod9** [comment]
Does this also fix https://github.com/dotnet/runtime/issues/41311, or is that a separate issue?

**Comment by Maoni0** [comment]
@mangod9 this is different. #41311 repros without g_GCShadow involved at all. that's why I asked to open a new issue.

**Comment by Maoni0** [comment]
please see [this comment](https://github.com/dotnet/runtime/issues/40058#issuecomment-680320838).

**Comment by cshung** [comment]
@Maoni0, I saw your comment. I am wondering if it makes more sense for us to have this tactical fix now to unblock GCStress runs for the JIT team?

The inability of using the shadow heap with local GC was captured 4 years ago and nobody worked on it. That probably implies nobody really needs the combination. That should not be a priority now.

**Comment by Maoni0** [comment]
well, I didn't think it was going to be much work, that's why I suggested it. but if you believe it's a fair amount of work, go ahead and check this in but I would not remove wbs_GCShadow.

**Comment by Maoni0** [comment]
and I would add a comment to indicate there's a TODO which is why wbs_GCShadow is still there.

**Comment by xiangzhai** [comment]
Hi @cshung 

```
diff --git a/src/coreclr/src/vm/arm64/asmhelpers.S b/src/coreclr/src/vm/arm64/asmhelpers.S
index a8b0a7c..f76a748 100644
--- a/src/coreclr/src/vm/arm64/asmhelpers.S
+++ b/src/coreclr/src/vm/arm64/asmhelpers.S
@@ -375,7 +375,7 @@ WRITE_BARRIER_ENTRY JIT_WriteBarrier
     // Update GC Shadow Heap
 
     // Do not perform the work if g_GCShadow is 0
-    ldr  x12, LOCAL_LABEL(wbs_GCShadow)
+    PREPARE_EXTERNAL_VAR g_GCShadow, x12
     cbz  x12, LOCAL_LABEL(ShadowUpdateDisabled)
 
     // need temporary register. Save before using.
```

But still repos:

```
./build-runtime.sh

export CORE_LIBRARIES=/home/zhaixiang/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.8.20361.2
export COMPlus_HeapVerify=2

gdb -ex=r --args /home/zhaixiang/runtime/artifacts/bin/coreclr/Linux.arm64.Debug/corerun /home/zhaixiang/coreclr-mips64-dev/bin/tests/Linux.arm64.Debug/GC/API/NoGCRegion/NoGC/NoGC.exe



Assert failure(PID 2648 [0x00000a58], Thread: 2648 [0x0a58]): !"Pointer updated without using write barrier"
    File: /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp Line: 39195
    Image: /home/zhaixiang/runtime/artifacts/bin/coreclr/Linux.arm64.Debug/corerun


Thread 1 "corerun" received signal SIGTRAP, Trace/breakpoint trap.
DBG_DebugBreak () at /home/zhaixiang/runtime/src/coreclr/src/pal/src/arch/arm64/debugbreak.S:7
7	    EMIT_BREAKPOINT
(gdb) bt
#0  DBG_DebugBreak () at /home/zhaixiang/runtime/src/coreclr/src/pal/src/arch/arm64/debugbreak.S:7
#1  0x0000007fb770cc9c in DebugBreak () at /home/zhaixiang/runtime/src/coreclr/src/pal/src/debug/debug.cpp:410
#2  0x0000007fb750bcd4 in DbgAssertDialog (szFile=0x7fb7879a20 "/home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp", iLine=39195, szExpr=0x7fb787e994 "!\"Pointer updated without using write barrier\"") at /home/zhaixiang/runtime/src/coreclr/src/utilcode/debug.cpp:697
#3  0x0000007fb74b7be4 in WKS::testGCShadow (ptr=0x7f0ffff478) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:39195
#4  WKS::testGCShadowHelper (x=0x7f0ffff470 "\260\274l>\177") at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:39212
#5  0x0000007fb7488050 in WKS::checkGCWriteBarrier () at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:39240
#6  0x0000007fb7487870 in WKS::gc_heap::garbage_collect (n=0) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:18093
#7  0x0000007fb74704e0 in WKS::GCHeap::GarbageCollectGeneration (this=0x55555c45f0, gen=0, reason=reason_alloc_soh) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:37747
#8  0x0000007fb74722cc in WKS::gc_heap::trigger_gc_for_alloc (gen_number=0, gr=reason_alloc_soh, msl=0x7fb7b7dce8 <WKS::gc_heap::more_space_lock_soh>, loh_p=false, take_state=WKS::mt_try_budget) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:13852
#9  0x0000007fb7473640 in WKS::gc_heap::try_allocate_more_space (acontext=0x55555ebd28, size=944, flags=0, gen_number=0) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:13975
#10 0x0000007fb7473924 in WKS::gc_heap::allocate_more_space (acontext=0x55555ebd28, size=944, flags=0, alloc_generation_number=0) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:14476
#11 0x0000007fb74b30a8 in WKS::gc_heap::allocate (jsize=939, acontext=0x55555ebd28, flags=0) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:14507
#12 WKS::GCHeap::Alloc (this=0x55555c45f0, context=0x55555ebd28, size=939, flags=0) at /home/zhaixiang/runtime/src/coreclr/src/gc/gc.cpp:36742
#13 0x0000007fb726ddd8 in Alloc (size=939, flags=GC_ALLOC_NO_FLAGS) at /home/zhaixiang/runtime/src/coreclr/src/vm/gchelpers.cpp:228
#14 0x0000007fb726c440 in AllocateSzArray (pArrayMT=0x7f3e8d9cc0, cElements=915, flags=GC_ALLOC_NO_FLAGS) at /home/zhaixiang/runtime/src/coreclr/src/vm/gchelpers.cpp:483
#15 0x0000007fb729e434 in JIT_NewArr1 (arrayMT=0x7f3e8d9cc0, size=915) at /home/zhaixiang/runtime/src/coreclr/src/vm/jithelpers.cpp:2723
#16 0x0000007fb729df04 in JIT_NewArr1VC_MP_FastPortable (arrayMT=0x7f3e8d9cc0, size=915) at /home/zhaixiang/runtime/src/coreclr/src/vm/jithelpers.cpp:2626
#17 0x0000007f3e5c9490 in ?? ()
#18 0x0000007f10015678 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
(gdb) x/22i 0x0000007f3e5c9490-44
   0x7f3e5c9464:	ldr	x2, [x2]
   0x7f3e5c9468:	ldr	x2, [x2,#72]
   0x7f3e5c946c:	ldr	x2, [x2,#56]
   0x7f3e5c9470:	blr	x2
   0x7f3e5c9474:	str	w0, [x29,#60]
   0x7f3e5c9478:	ldr	w1, [x29,#60]
   0x7f3e5c947c:	sxtw	x1, w1
   0x7f3e5c9480:	mov	x0, #0x9cc0                	// #40128
   0x7f3e5c9484:	movk	x0, #0x3e8d, lsl #16
   0x7f3e5c9488:	movk	x0, #0x7f, lsl #32
   0x7f3e5c948c:	bl	0x7f3e57dfc0
   0x7f3e5c9490:	str	x0, [x29,#48]
   0x7f3e5c9494:	ldr	x0, [x29,#48]
   0x7f3e5c9498:	str	x0, [x29,#72]
   0x7f3e5c949c:	ldr	w0, [x29,#84]
   0x7f3e5c94a0:	ldr	w1, [x29,#88]
   0x7f3e5c94a4:	cmp	w1, #0x0
   0x7f3e5c94a8:	b.ne	0x7f3e5c94b0
   0x7f3e5c94ac:	bl	0x7f3e57dfa0
   0x7f3e5c94b0:	cmn	w1, #0x1
   0x7f3e5c94b4:	b.ne	0x7f3e5c94c8
   0x7f3e5c94b8:	cmn	w0, w0
```

Please point out my fault.

Thanks,
Leslie Zhai

**Comment by cshung** [comment]
> But still repos:

@xiangzhai, thanks for testing out my fix proactively and I am sorry that it still repros for you.

All I have right now is a Windows arm64 machine, I have run the NoGC test in a loop for 4 hours on that machine and I couldn't repro it with the fix.

There could be a Linux specific issue or just pure luck that I didn't hit it in a few hours. How easy is it for you to repro the assertion after taking my fix?

**Comment by xiangzhai** [comment]
Hi @cshung 

You are welcome!  I am also debugging the issue for MIPS64 :)

> There could be a Linux specific issue or just pure luck that I didn't hit it in a few hours. How easy is it for you to repro the assertion after taking my fix?

100 failures:

```
./build-runtime.sh

export CORE_LIBRARIES=/home/zhaixiang/runtime/.dotnet/shared/Microsoft.NETCore.App/5.0.0-preview.8.20361.2
export COMPlus_HeapVerify=2

for i in {1..100}
do
/home/zhaixiang/runtime/artifacts/bin/coreclr/Linux.arm64.Debug/corerun /home/zhaixiang/coreclr-mips64-dev/bin/tests/Linux.arm64.Debug/GC/API/NoGCRegion/NoGC/NoGC.exe
done
```

The ARM64 test machine: 

```
$ cat /proc/meminfo 
MemTotal:        4034620 kB
MemFree:          986556 kB
MemAvailable:    3477080 kB
Buffers:          135784 kB
Cached:          2211472 kB
SwapCached:        22684 kB


$ free -lh
              total        used        free      shared  buff/cache   available
Mem:           3.8G        470M        963M          9M        2.4G        3.3G
Low:           3.8G        2.9G        963M
High:            0B          0B          0B
Swap:          8.0G        905M        7.1G
```

Thanks,
Leslie Zhai

File: src/coreclr/src/vm/arm64/asmhelpers.S (hunk 1 above)
Comment (cshung) [reply src/coreclr/src/vm/arm64/asmhelpers.S:378]: Actually, I think we will leave it and add a comment on it instead.

**Comment by xiangzhai** [comment]
Linux ARM64 PASSED:  https://github.com/dotnet/runtime/pull/41375

**Pull request closed without merging**